### PR TITLE
Server Connector

### PR DESCRIPTION
**Pull request by KDharmarajanDev**
This pull request includes:
- Autoconnecting to the server (IP still needs to be changed to actual server ip)
- Single service publishing to the roslibpy side
- MAVROS launching
- Namespace Changing (/drone_1/.....)

This DOES NOT Include:
- Safety mechanisms
- All essential MAVROS services/topics published to roslibpy
- Tested simulation + server control

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,55 @@
...
+    1. For ArduPilot, run the following: `git clone https://github.com/ArduPilot/ardupilot.git`
+    Change directory into ardupilot via `cd ardupilot`. Run `git submodule update --init --recursive`. 
+    From inside the ardupilot directory, run `Tools/environment_install/install-prereqs-ubuntu.sh -y`. Finally,
+    run `. ~/.profile`. For more information, visit https://ardupilot.org/dev/docs/building-setup-linux.html#building-setup-linux.
+    
+    2. For PX4, run the following: `git clone https://github.com/PX4/PX4-Autopilot.git --recursive`.
+    Change directory into PX4-Ardupilot via `cd PX4-Ardupilot`. Run `bash ./Tools/setup/ubuntu.sh`.
+    For more information, visit https://docs.px4.io/master/en/dev_setup/dev_env_linux_ubuntu.html.
+    
+2. Please install Gazebo by following [this link](http://gazebosim.org/tutorials?tut=install_ubuntu&cat=install)
```
Comment (erwang01) [inline README.md]: Some ROS installations already have Gazebo installed. Not sure if this is worth mentioning.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,55 @@
...
+
+2. Install geographiclib and mavlink
+    ```
+    sudo apt-get install geographiclib-* ros-melodic-geographic-*  
+    sudo apt-get install ros-melodic-mavlink
+    ```
+   
+3. Clone this repository by running: `git clone https://github.com/immersive-command-system/drone-mavros.git --recursive`.
+
+4. Once in the directory of the repository, run `catkin build`
```
Comment (erwang01) [inline README.md]: Which directory?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,55 @@
...
+
+1. Please install ROS by following [this link](http://wiki.ros.org/melodic/Installation)
+
+2. Install geographiclib and mavlink
+    ```
+    sudo apt-get install geographiclib-* ros-melodic-geographic-*  
+    sudo apt-get install ros-melodic-mavlink
+    ```
+   
+3. Clone this repository by running: `git clone https://github.com/immersive-command-system/drone-mavros.git --recursive`.
```
Comment (erwang01) [inline README.md]: Does this repo need to be cloned into a catkin workspace?

File: README.md (hunk 1 above)
Comment (KDharmarajanDev) [reply README.md]: I just included it in case someone didn't have Gazebo.

File: README.md (hunk 3 above)
Comment (KDharmarajanDev) [reply README.md]: This itself can be a catkin workspace via the src folder.

File: README.md (hunk 2 above)
Comment (KDharmarajanDev) [reply README.md]: drone-mavros. I should specify it.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,67 @@
...
+1. `cd drone-mavros`
+2. `git submodule update --init --recursive`
+3. `catkin build`
+4. `source devel/setup.bash`
+
+`catkin: command not found`
+1. Try `sudo apt-get install ros-$ROS_DISTRO-catkin python-catkin-tools`
+    
+When `libmavconn` package, `CMake Error` finding package configuration file provided by `mavlink`.
+1. `sudo apt-get install ros-kinetic-mavros ros-kinetic-mavros-extras`
```
Comment (erwang01) [inline README.md:68]: If we do this already do we even need the mavros submodule?

Comment (erwang01) [reply README.md:68]: Because mavros and mavros_extras take forever to compute.

File: src/server_connector/src/connect_to_server.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+rospy.loginfo('Starting!')
+
+server_ip = rospy.get_param('server_ip', '54.161.15.175')
+server_port = 9090
+server_connection = roslibpy.Ros(host=server_ip, port=server_port)
+server_connection.run(timeout=10)
+rospy.loginfo('Attempting to connect to: ' + server_ip)
+result = None
+if server_connection.is_connected:
+    result = register_drone(server_connection, 'hexacopter2')
```
Comment (erwang01) [inline src/server_connector/src/connect_to_server.py]: Can we have this placed in a constants file or some sort of environment variable?

File: src/server_connector/src/connect_to_server.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    uuid = roslaunch.rlutil.get_or_generate_uuid(None, False)
+    roslaunch.configure_logging(uuid)
+    roslaunch_file = [(new_path,
+                       ['namespace:=' + namespace])]
+    parent = roslaunch.parent.ROSLaunchParent(uuid, roslaunch_file)
+    parent.start()
+
+rospy.loginfo('Starting!')
+
+server_ip = rospy.get_param('server_ip', '54.161.15.175')
```
Comment (erwang01) [inline src/server_connector/src/connect_to_server.py:8]: Can we have the IP as an environment var / constant?

File: src/server_connector/src/connect_to_server.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env python
+import rospy
+from connection_manager import ConnectionManager
+
+rospy.init_node('server_connector')
+rospy.loginfo('Starting!')
+
+server_ip = rospy.get_param('server_ip', '54.161.15.175')
+server_port = 9090
+connection_manager = ConnectionManager(server_ip, server_port, 'hexacopter')
```
Comment (erwang01) [inline src/server_connector/src/connect_to_server.py:10]: Can we have the Name also as an environment var/constant?

**Review by erwang01 — APPROVED** [review]
LGTM builds properly now. Have not done full system integration test yet.

**Comment by erwang01** [comment]
Integration test complete. Everything works.

Do we know how to shutdown the drone tho? Is there a hot key for that or smtg?

File: src/server_connector/src/connect_to_server.py (hunk 6 above)
Comment (erwang01) [reply src/server_connector/src/connect_to_server.py:8]: or via launch file.

either way document in README

**Pull request merged**